### DISCUSSION TITLE
I keep getting an error about argmax and empty series when using a custom indicator function

### DISCUSSION THREAD

**Discussion by Precisionguided** [kernc/backtesting.py#433] [General]
Thanks for the great tool you've provided! I've been able to backtest strategies based on the talib indicators, but recently, I've been trying to convert one of my more complex tradingview strategies into backtesting.py but I'm having an issue with uncertainty about how data is passed to the functions. Pine script evaluates everything one candle at a time, but when I wrap an indicator it seems like it evaluates everything at once, which doesn't allow me to reference x[-1], x[-2] to perform operations on them on a rolling basis. I've tried to figure it out by looking at backtesting.Strategy but I'm still having a bit of difficulty figuring it out and I've tried using .rolling but it won't let me perform operations on variables that use rolling.. I was wondering if you could provide some direction.

Am I supposed to take a for loop and process it then output everything in an array, rather than doing everything as a stream?

**Comment by kernc** [comment]
It's way easier to talk about code with some actual sample code.

**Reply by Precisionguided** [reply]
Thanks for the reply. I was able to get it running, but I was wondering if there's a neater way to output custom indicators than running for loops, like so:

```py
    def Cond_EMA(x,cond): 
    
        np.nan_to_num(x,copy=False)
    
        ema_val=np.array([])
        
        for i in range(0,-len(x),-1):
    
            if cond==1:
    
                if not ema_val.size>0:
    
                    ema_val=np.append(ema_val,x[-i],axis=None)
                
                else:
                            
                    ema_val=np.append(ema_val,((x[-i]-ema_val[-1])*((2/(rng_per+1)))+ema_val[-1]),axis=None)
                
        return pd.Series(ema_val)
```

**Reply by kernc** [reply]
Firstly, your `range(0,-len(x),-1)` is quite weird. This indexes:
```py
>>> x = np.r_[1, 2, 3, 4]
>>> x[list(range(0, -len(x), -1))]
array([1, 4, 3, 2])
```
which _might_ be what you want.

Otherwise, yes, you need the loop if you want to [refer to previously computed value](https://stackoverflow.com/questions/34855859/is-there-a-way-in-pandas-to-use-previous-row-value-in-dataframe-apply-when-previ).
Also look into [pandas DataFrame EMA computation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html).

**Reply by Precisionguided** [reply]
Yeah the index is weird because I wanted to work with -i and -i-1 as indexes. Your EMA made my code a lot neater, thanks.

Lastly, I was wondering if there is a way to make optimize use all my cores? So far it seems to be using very little of my cpu power. The backtest is like this as well but I guess that's not parallelized and for the backtest it's pretty quick anyway. Thanks in advance.

**Reply by kernc** [reply]
> Lastly, I was wondering if there is a way to make optimize use all my cores?

[Optimization `method="grid"`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest.optimize) parallelizes by default, but it's broken on Windos as the [TODO note says](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest.optimize). You're very welcome to investigate more proper support on Windos:
https://github.com/kernc/backtesting.py/blob/4f21a652d5292f93d5a717cc25f068fd2c35dcc2/backtesting/backtesting.py#L1360-L1379

**Reply by Precisionguided** [reply]
I'd love to but I'm having to study how as I go along. It's a little bit beyond my programming ability at the moment, though. I tried running optimize on my WSL ubuntu install but it looks like it's only using one cpu also. I also didn't get the warning about setting multiprocessing start method to 'fork'. Is there something else I could be doing wrong? Or could it be because I'm using WSL? Although I've had other programs utilize all cores through that.

**Reply by Precisionguided** [reply]
Not sure how you'd integrate it since, like I said, I'm not really well-versed in programming, but I was able to utilize all my cores for optimization with hyperactive, which has similar optimization algorithms to skopt. The ff is an example using particle swarm where n_jobs is the number of jobs to do concurrently. The jobs don't communicate with each other, though.

```

    def run_strat(params):
    
        bt_cb = Backtest(eth_df_hour.loc['2020-06-01':'2021-04-01'], SMAcross, cash=1000000, commission=.0012,hedging=True)

        stats=bt_cb.run(**params)

        bt_cb

        return stats['Sortino Ratio']

    from hyperactive import Hyperactive
    from hyperactive import ParticleSwarmOptimizer
    from hyperactive.dashboards import ProgressBoard

    search_space = {
        'rng_scale':list([1,2,3]),
        'smooth_per':list(range(3,13,3)),
    }

    iterations=1000
    
    hyper=Hyperactive
    
    hyper.add_search(run_strat,search_space,optimizer=ParticleSwarmOptimizer(),n_iter= \ 
    iterations,n_jobs=16,progress_board=progress_board) 

    hyper.run()
    res=hyper.results(run_strat)
 ```

**Reply by kernc** [reply]
> on my WSL ubuntu install but it looks like it's only using one cpu also

Can you share a toy code example you're trying with? The [example notebook](https://kernc.github.io/backtesting.py/doc/examples/Parameter%20Heatmap%20&%20Optimization.html) ([colab](https://colab.research.google.com/github/kernc/backtesting.py/blob/master/doc/examples/Parameter%20Heatmap%20%26%20Optimization.ipynb#scrollTo=Vh6TONA_dxRS&line=16&uniqifier=1)) first optimization should paralellize. If you `pip install tqdm`, you can see the progress bar in action.
Alternatively, might have something to do with job batching:
https://github.com/kernc/backtesting.py/blame/4f21a652d5292f93d5a717cc25f068fd2c35dcc2/backtesting/backtesting.py#L1347-L1350

**Reply by Precisionguided** [reply]
That notebook worked for me. Will do some more tests to figure out why it's not working on my bigger test. Last question: is there a way to buy with a percentage of over 100 percent? As in with a percentage of equity, margined, but without specifying a particular number of coins to buy each time?

**Reply by kernc** [reply]
> a percentage of equity, margined

You e.g. [`Backtest(..., margin=.1)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest) for a 10x leverage.
